_posts/2024-10-24-why-new-rtos.markdown

```diff
@@ -0,0 +1,212 @@
+If there is a bounds error in addressing into A then the component may accidentally overwrite B.
+In contrast, in a CHERI system, you would pass bounded pointers (CHERI capabilities) to A and B.
+Any operation that attempts to access A will never accidentally access B.
+This intentionality (you can access an object only when you intend to access *that object*) is core to CHERI.
+
+The CHERIoT security model was designed to support language-friendly abstractions.
+You don't share objects in CHERIoT RTOS by creating shared regions of memory and populating them, you share objects by passing pointers as arguments to functions.
+This comes with a rich hardware-enforced permission model that lets you enforce properties such as shallow or deep immutability or no-capture guarantees on pointers.
+
+For example, the TLS compartment in the network stack maintains a per-connection ring buffer that it uses for incoming and outgoing messages.
+When it wishes to receive some data, it creates a bounded pointer to a region in that buffer that has only store permission.
+The TLS compartment then passes this to the TCP/IP compartment's receive function.
+When this call returns, the TLS compartment knows that this buffer is safe from time-of-check-to-time-of-use problems because the TCP/IP compartment cannot have captured it because the pointer does not have global permission.
+It knows that no data outside of that buffer can have been written, because the pointer is bounded.
+It knows that data left over in that buffer cannot have been read (because the pointer does not have read permission) and so it is safe from information disclosure.
+All of this is visible in the source code and is enforced with CHERI.
+
+It might be possible to get the same guarantees from an MPU by invoking a kernel to set up an MPU region with store-only permission covering an object owned by the TLS stack, but this would look very different.
+Trying to provide the same programmer model for both CHERI and MPU-based systems (or even those without an MPU) would require serious compromises.
+
+MMUs were not designed to provide abstractions for normal programmers.
+They were created to provide operating-system abstractions such as swapping and the ability to isolate processes or virtual machines from each other.
+MPUs were designed to allow an RTOS kernel to protect itself from untrusted tasks.
+They were both designed as tools for use by operating systems.
+
+CHERI, in contrast, was designed to provide a protection model that could be directly exposed into C and higher-level programming languages.
+It is designed to protect things that are visible in a language abstract machine as objects, not memory ranges.
+It is designed to protect them by putting the permissions (and bounds and integrity guarantees) in things that are exposed to programming languages as pointers, not as entries in an indirection table.
```

Comment:
I'm not elated with the repeated "It is designed to protect", so perhaps something like this, though I can't say as it sparks joy, exactly, either.

```suggestion
It protects things that are visible in a language abstract machine as objects, not their underlying memory ranges.
It does so by conveying rights (permissions and bounds and integrity guarantees) using things that are exposed to programming languages as pointers, not as entries in an indirection table.
```